lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,746 @@
+        } catch (error) {
+          process.nextTick(() => writable.destroy(error));
+        }
+      }
+
+      PromisePrototypeThen(
+        writer.ready,
+        () => {
+          return PromisePrototypeThen(
+            PromiseAll(
+              ArrayPrototypeMap(
+                chunks,
+                (chunk) => writer.write(chunk))),
+            done,
+            done);
+        },
+        done);
+    },
+
+    write(chunk, encoding, callback) {
+      if (closed) {
+        callback(new ERR_INVALID_STATE('The WritableStream is closed'));
+        return;
+      }
+      if (typeof chunk === 'string')
+        chunk = new Uint8Array(Buffer.from(chunk, encoding));
+
+      function done(error) {
+        try {
+          callback(error);
+        } catch (error) {
+          process.nextTick(() => writable.destroy(error));
+        }
+      }
+
+      PromisePrototypeThen(
+        writer.ready,
+        () => {
+          return PromisePrototypeThen(
+            writer.write(chunk),
+            done,
+            done);
+        },
+        done);
+    },
+
+    destroy(error, callback) {
+      function done() {
+        try {
+          callback(error);
+        } catch (error) {
+          process.nextTick(() => { throw error; });
+        }
+      }
+
+      if (!closed) {
+        PromisePrototypeThen(
+          writer.abort(error),
+          done,
+          done);
+      }
+    },
+
+    final(callback) {
+      function done(error) {
+        try {
+          callback(error);
+        } catch (error) {
+          process.nextTick(() => writable.destroy(error));
+        }
+      }
+
+      if (!closed) {
+        PromisePrototypeThen(
+          writer.close(),
+          done,
+          done);
+      }
+    },
+  });
+
+  PromisePrototypeThen(
+    writer.closed,
+    () => {
+      closed = true;
+      if (!writable.writableEnded)
+        writable.end();
+    },
+    (error) => {
+      closed = true;
+      if (!writable.destroyed)
+        writable.destroy(error);
+    });
+
+  return writable;
+}
+
+function newStreamReadableFromReadableStream(readableStream, options = {}) {
+  if (!isReadableStream(readableStream)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'readableStream',
+      'ReadableStream',
+      readableStream);
+  }
+
+  validateObject(options, 'options');
+  const {
+    highWaterMark,
+    encoding,
+    objectMode = false,
+    signal,
+  } = options;
+
+  if (encoding !== undefined && !Buffer.isEncoding(encoding))
+    throw new ERR_INVALID_ARG_VALUE(encoding, 'options.encoding');
+  validateBoolean(objectMode, 'options.objectMode');
+  if (signal !== undefined)
+    validateAbortSignal(signal, 'options.signal');
+
+  const reader = readableStream.getReader();
+  let closed = false;
+
+  function doRead() {
+    PromisePrototypeThen(
+      reader.read(),
+      (chunk) => {
+        // We can keep reading while readable.push returns true.
+        if (readable.push(chunk))
```

Comment:
Does `Readable` guarantee that if `readable.push()` returns `true`, then `readable._read()` will **not** be called? Otherwise, we may risk starting too many reads.